File: connect.py
#!/usr/bin/env python3
"""connect.py — super-lite MCP client for an agent-integrations relay session.

Connects to a session-based MCP relay (the protocol shipped by
@particle-academy/agent-integrations): POST JSON-RPC to /inbox, read the host's
responses from a server-sent-events /events stream. Drives any app that mounts
the relay + a MicroMcpServer (e.g. the Fancy UI Agent Playground).

Deps: Python 3.8+ stdlib only.

Usage:
  python3 connect.py <url> tools                  # list the host's tools
  python3 connect.py <url> call <name> ['<json>']  # call a tool (args default {})
  python3 connect.py <url> send '<jsonrpc>'        # send a raw JSON-RPC frame
  python3 connect.py <url> watch                   # stream every frame from host

<url> is whatever connection URL you were handed; the token (the "inline key")
may be in it (?token=… or ?key=…) or supplied via MCP_TOKEN. Forms:
  https://host/agent-playground?session=ABC&token=XYZ
  https://host/whiteboard-share/ABC?token=XYZ
  https://host/whiteboard-share/ABC/inbox?token=XYZ

Env: MCP_TOKEN, MCP_RELAY_PATH (default whiteboard-share), MCP_INSECURE (skip TLS).
"""
import json
import os
import ssl
import sys
import threading
import time
import urllib.parse
import urllib.request

RELAY_PATH = os.environ.get("MCP_RELAY_PATH", "whiteboard-share")
INSECURE = bool(os.environ.get("MCP_INSECURE"))
SSL_CTX = ssl._create_unverified_context() if INSECURE else None


def die(msg):
    sys.stderr.write("error: " + msg + "\n")
    sys.exit(1)


def endpoints(url):
    """Parse a connection URL into (inbox, events) relay endpoints."""
    u = urllib.parse.urlparse(url)
    q = urllib.parse.parse_qs(u.query)
    token = os.environ.get("MCP_TOKEN") or (q.get("token") or q.get("key") or [""])[0]
    origin = "%s://%s" % (u.scheme, u.netloc)
    if "session" in q:
        session = q["session"][0]
        base = "%s/%s" % (origin, RELAY_PATH.strip("/"))
    else:
        segs = [s for s in u.path.split("/") if s]
        if segs and segs[-1] in ("inbox", "events", "outbox"):
            segs = segs[:-1]
        session = segs[-1] if segs else ""
        base = origin + "".join("/" + s for s in segs[:-1])
    if not token:
        die("no token in URL and MCP_TOKEN unset")
    if not session:
        die("could not determine session from URL")
    inbox = "%s/%s/inbox?token=%s" % (base, session, token)
    events = "%s/%s/events?token=%s&direction=outbound" % (base, session, token)
    return inbox, events


class Relay:
    def __init__(self, inbox, events):
        self.inbox = inbox
        self.events = events
        self.lock = threading.Lock()
        self.responses = {}
        self.watch = False

    def _read_sse(self):
        req = urllib.request.Request(self.events, headers={"accept": "text/event-stream"})
        with urllib.request.urlopen(req, context=SSL_CTX) as resp:
            for raw in resp:
                line = raw.decode("utf-8", "ignore").rstrip("\r\n")
                if not line.startswith("data: "):
                    continue
                try:
                    frame = json.loads(line[6:])
                except ValueError:
                    continue
                if self.watch:
                    print(json.dumps(frame))
                    sys.stdout.flush()
                elif "id" in frame:
                    with self.lock:
                        self.responses[frame["id"]] = frame

    def start(self):
        threading.Thread(target=self._read_sse, daemon=True).start()
        time.sleep(1.0)  # let the subscription register (pings host: peer_joined)

    def post(self, frame):
        data = json.dumps(frame).encode("utf-8")
        req = urllib.request.Request(
            self.inbox, data=data, method="POST",
            headers={"content-type": "application/json"},
        )
        urllib.request.urlopen(req, context=SSL_CTX).read()

    def await_id(self, frame_id, timeout=15.0):
        end = time.time() + timeout
        while time.time() < end:
            with self.lock:
                if frame_id in self.responses:
                    return self.responses[frame_id]
            time.sleep(0.1)
        return None


def main():
    args = sys.argv[1:]
    if len(args) < 2:
        sys.stderr.write(__doc__)
        sys.exit(2)
    url, cmd, rest = args[0], args[1], args[2:]
    inbox, events = endpoints(url)
    relay = Relay(inbox, events)

    if cmd == "watch":
        relay.watch = True
        sys.stderr.write("# watching (Ctrl-C to stop)\n")
        relay._read_sse()
        return

    relay.start()
    relay.post({"jsonrpc": "2.0", "id": 1, "method": "initialize",
                "params": {"protocolVersion": "2025-06-18", "capabilities": {},
                           "clientInfo": {"name": "connect.py", "version": "1"}}})
    if relay.await_id(1) is None:
        die("no response — is the session live and the host connected?")
    relay.post({"jsonrpc": "2.0", "method": "notifications/initialized"})

    if cmd == "tools":
        relay.post({"jsonrpc": "2.0", "id": 2, "method": "tools/list"})
        result = relay.await_id(2)
    elif cmd == "call":
        if not rest:
            die("usage: connect.py <url> call <name> ['<json-args>']")
        name = rest[0]
        arguments = json.loads(rest[1]) if len(rest) > 1 else {}
        relay.post({"jsonrpc": "2.0", "id": 3, "method": "tools/call",
                    "params": {"name": name, "arguments": arguments}})
        result = relay.await_id(3)
    elif cmd == "send":
        if not rest:
            die("usage: connect.py <url> send '<jsonrpc-frame>'")
        frame = json.loads(rest[0])
        relay.post(frame)
        result = relay.await_id(frame["id"]) if "id" in frame else None
    else:
        die("unknown command: %s (use tools|call|send|watch)" % cmd)

    print(json.dumps(result, indent=2) if result is not None else "(no response)")


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        pass

File: connect.ts
#!/usr/bin/env -S npx tsx
//
// connect.ts — super-lite MCP client for an agent-integrations relay session.
//
// Connects to a session-based MCP relay (the protocol shipped by
// @particle-academy/agent-integrations): POST JSON-RPC to /inbox, read the
// host's responses from a server-sent-events /events stream. Drives any app
// that mounts the relay + a MicroMcpServer (e.g. the Fancy UI Agent Playground).
//
// No npm deps — uses the runtime's built-in fetch. Run with any of:
//   bun connect.ts <url> ...
//   npx tsx connect.ts <url> ...
//   deno run -A connect.ts <url> ...
//   node --experimental-strip-types connect.ts <url> ...   (Node 22.6+)
//
// Usage:
//   connect.ts <url> tools                  # list the host's tools
//   connect.ts <url> call <name> ['<json>']  # call a tool (args default {})
//   connect.ts <url> send '<jsonrpc>'        # send a raw JSON-RPC frame
//   connect.ts <url> watch                   # stream every frame from the host
//
// <url> is whatever connection URL you were handed; the token (the "inline
// key") may be in it (?token=… / ?key=…) or supplied via MCP_TOKEN.
//
// Env: MCP_TOKEN, MCP_RELAY_PATH (default whiteboard-share), MCP_INSECURE (skip TLS).

// Skip TLS verification for self-signed certs (local dev) — must run before fetch.
if (process.env.MCP_INSECURE) process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

const RELAY_PATH = process.env.MCP_RELAY_PATH || "whiteboard-share";

type Frame = { jsonrpc: string; id?: number | string; method?: string; result?: unknown; error?: unknown; params?: unknown };

function die(msg: string): never {
  process.stderr.write("error: " + msg + "\n");
  process.exit(1);
}

function endpoints(rawUrl: string): { inbox: string; events: string } {
  const u = new URL(rawUrl);
  const token = process.env.MCP_TOKEN || u.searchParams.get("token") || u.searchParams.get("key") || "";
  const origin = u.origin;
  let session = "";
  let base = "";
  const qSession = u.searchParams.get("session");
  if (qSession) {
    session = qSession;
    base = `${origin}/${RELAY_PATH.replace(/^\/|\/$/g, "")}`;
  } else {
    let segs = u.pathname.split("/").filter(Boolean);
    if (segs.length && ["inbox", "events", "outbox"].includes(segs[segs.length - 1])) segs = segs.slice(0, -1);
    session = segs[segs.length - 1] || "";
    base = origin + segs.slice(0, -1).map((s) => "/" + s).join("");
  }
  if (!token) die("no token in URL and MCP_TOKEN unset");
  if (!session) die("could not determine session from URL");
  return {
    inbox: `${base}/${session}/inbox?token=${token}`,
    events: `${base}/${session}/events?token=${token}&direction=outbound`,
  };
}

async function main(): Promise<void> {
  const [url, cmd, ...rest] = process.argv.slice(2);
  if (!url || !cmd) {
    process.stderr.write("usage: connect.ts <url> {tools|call <name> [json]|send <frame>|watch}\n");
    process.exit(2);
  }
  const { inbox, events } = endpoints(url);

  const pending = new Map<number | string, (f: Frame) => void>();
  const watching = cmd === "watch";

  // Background SSE reader.
  const ac = new AbortController();
  const readSSE = async (): Promise<void> => {
    const res = await fetch(events, { signal: ac.signal, headers: { accept: "text/event-stream" } });
    if (!res.body) die("no SSE stream");
    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buf = "";
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buf += decoder.decode(value, { stream: true });
      let i: number;
      while ((i = buf.indexOf("\n\n")) >= 0) {
        const block = buf.slice(0, i);
        buf = buf.slice(i + 2);
        const line = block.split("\n").find((l) => l.startsWith("data: "));
        if (!line) continue;
        let frame: Frame;
        try {
          frame = JSON.parse(line.slice(6));
        } catch {
          continue;
        }
        if (watching) {
          console.log(JSON.stringify(frame));
        } else if (frame.id != null && pending.has(frame.id)) {
          pending.get(frame.id)!(frame);
          pending.delete(frame.id);
        }
      }
    }
  };

  if (watching) {
    process.stderr.write("# watching (Ctrl-C to stop)\n");
    await readSSE();
    return;
  }

  readSSE().catch(() => {});
  const post = (frame: Frame) =>
    fetch(inbox, { method: "POST", headers: { "content-type": "application/json" }, body: JSON.stringify(frame) });
  const awaitId = (id: number | string, timeout = 15000): Promise<Frame | null> =>
    new Promise((resolve) => {
      pending.set(id, resolve);
      setTimeout(() => {
        if (pending.delete(id)) resolve(null);
      }, timeout);
    });

  await new Promise((r) => setTimeout(r, 1000)); // let SSE subscribe (pings host)
  await post({ jsonrpc: "2.0", id: 1, method: "initialize", params: { protocolVersion: "2025-06-18", capabilities: {}, clientInfo: { name: "connect.ts", version: "1" } } });
  if ((await awaitId(1)) === null) die("no response — is the session live and the host connected?");
  await post({ jsonrpc: "2.0", method: "notifications/initialized" });

  let result: Frame | null;
  if (cmd === "tools") {
    await post({ jsonrpc: "2.0", id: 2, method: "tools/list" });
    result = await awaitId(2);
  } else if (cmd === "call") {
    const name = rest[0];
    if (!name) die("usage: connect.ts <url> call <name> ['<json-args>']");
    const args = rest[1] ? JSON.parse(rest[1]) : {};
    await post({ jsonrpc: "2.0", id: 3, method: "tools/call", params: { name, arguments: args } });
    result = await awaitId(3);
  } else if (cmd === "send") {
    if (!rest[0]) die("usage: connect.ts <url> send '<jsonrpc-frame>'");
    const frame: Frame = JSON.parse(rest[0]);
    await post(frame);
    result = frame.id != null ? await awaitId(frame.id) : null;
  } else {
    die(`unknown command: ${cmd} (use tools|call|send|watch)`);
  }

  console.log(result != null ? JSON.stringify(result, null, 2) : "(no response)");
  ac.abort();
  process.exit(0);
}

main().catch((e) => die(e instanceof Error ? e.message : String(e)));
